pkg/mimir/modules.go

```diff
@@ -719,7 +719,16 @@ func (t *Mimir) initQueryFrontendTopicOffsetsReader() (services.Service, error)
 		return nil, err
 	}
 
-	t.QueryFrontendTopicOffsetsReader = ingest.NewTopicOffsetsReader(kafkaClient, t.Cfg.IngestStorage.KafkaConfig.Topic, t.Cfg.IngestStorage.KafkaConfig.LastProducedOffsetPollInterval, t.Registerer, util_log.Logger)
+	// The Kafka partitions may have been pre-provisioned, so the existing partitions in Kafka may be way more
+	// than the actual number of partitions we use. To improve performance, we only look up the actual partitions
+	// we're currently using in Mimir. We include all partition states because ACTIVE and INACTIVE partitions
+	// must be queried, and PENDING partitions may switch to ACTIVE between when the query-frontend fetch the offsets
+	// and the querier builds the replicaset of partitions to query.
+	getPartitionIDs := func() []int32 {
+		return t.IngesterPartitionRingWatcher.PartitionRing().PartitionIDs()
+	}
```

Comment:
To double-check I got it: since `memberlist` itself — implementation underneath the ring — is eventually consistent. Could there be any problems when the state any partition didn't yet fully propagate through the frontend's peers?

Reply:
Yes, absolutely. The partitions discovery done by query-frontend is just a best effort. The ultimate discovery is done by querier (which BTW could suffer the same issue too, and there's not much we can do about it today). If a strongly consistency request hits an ingester and the offset is missing in the request, then the ingester will fetch the last produced offset again for their partition in order to enforce it. This is an edge case that we don't expect to happen in the real world (excluding the few seconds after we add a new partition) and there's an alert firing if it happens for few minutes consecutively.